src/VisualStudio/Core/Def/ProjectSystem/MetadataReferences/VisualStudioMetadataReferenceManager.cs

```diff
@@ -328,14 +318,23 @@ bool TryGetFileMappingFromMetadataImporter(FileKey fileKey, [NotNullWhen(true)]
 
             // Suppression should be removed or addressed https://github.com/dotnet/roslyn/issues/41636
             var moduleFileKey = FileKey.Create(PathUtilities.CombineAbsoluteAndRelativePaths(assemblyDir, moduleName)!);
-            var metadata = moduleMetadataFactory(moduleFileKey);
+            var (metadata, metadataStorageHandle) = moduleMetadataFactory(moduleFileKey);
 
             moduleBuilder.Add(metadata);
+            storageHandles.Add(metadataStorageHandle);
         }
 
         if (moduleBuilder.Count == 0)
             moduleBuilder.Add(manifestModule);
 
-        return AssemblyMetadata.Create(moduleBuilder.ToImmutable());
+        var result = AssemblyMetadata.Create(moduleBuilder.ToImmutable());
+
+        // If we got any null handles, then we weren't able to map this whole assembly into memory mapped files. So we
+        // can't use those to transfer over the data efficiently to the OOP process.  In that case, we don't store the
+        // handles at all.
+        if (storageHandles.Count > 0 && storageHandles.All(h => h != null))
```

Comment:
>storageHandles.Count > 0 [](http://example.com/codeflow?start=12&length=24)

You add to this before the loop, can never be zero, right? Also, you know the size up front, can use your fixed size builder here too

Reply:
true.